### ISSUE TITLE
IO Thread getting into state where it is never ready to write

### ISSUE THREAD

**Issue by youngm** [hazelcast/hazelcast#5209]
Hazelcast 3.4.2

This issue is the result of my analysis of this Mailing list Thread: https://groups.google.com/d/msg/hazelcast/ksUBy4bYcSc/Qz1HDWjZb90J

When running stress tests on my 10 node cluster I've found that occasionally the WriteHandler for a IO Thread can get into a state where it is never ready because the channel isn't writing anything.

The state is that WriteHandler.ready == false and everytime AbstractIOSelector:124 is invoked:
```
selectedKeyCount = selector.select(waitTime);
```
The value of selectedKeyCount is ALWAYS 0.

The result is that this WriteHandler never responds to any requests from a certain node (the node associated with the handler).

It is difficult to get into this state so I cannot provide a simple test case duplicating the issue.  But I do have a theory of how this could happen.

In WriteHandler.handle():148 the code is not actually registering for a Write (line: 153) until after it has written to the buffer (line: 148).  Hence if it were possible for for the channel to process the buffer and flag it as done writing before registerWrite() is invoked then I think the WriteHandler would end up in the state I have described.  Never ready to write and the write Selector never saying it has finished writing anything.

Thoughts?  I can regularly duplicate this issue in my environment given an hour or 2 of giving it load.

Mike

**Comment by Donnerbart** [comment]
@mdogan @pveentjer Could you please have a look at this?

**Comment by noctarius** [comment]
Super nice work @youngm, thanks for you effort in digging into this :)

**Comment by sancar** [comment]
Always getting 0 from select should cause more Node communication to fail. There are 3 threads doing write. In 10 node cluster, 3 writes should be stuck after that point. It may be the case that if you wait a little bit longer after first timeout you will see a couple more timeouts in other nodes. 

And thanks for investigating this much, but a reproducible test code really help us a lot, if you can. Or can you rephrase your fail to theory again ? I could not follow your reasoning very well.

**Comment by youngm** [comment]
@sancar Dang.  OK, I did a little more studying of the code involved.  Let me try and rephrase.  In the debugging I did above I actually set my IO threads to 16 trying stuff out which created more of a 1-1 mapping between Thread and Selector which caused me to confuse my message a bit.  So, let me try again.

The way I understand it there is a WriteHandler for each TcpIpConnection.  In my situation the WriteHandler for one of the connections got into a state where ready is always FALSE.  Therefore never actually writing anything to the connection.  I believe it stays in this state because it is in a state that its SelectionKey is never getting triggered for selection in any IO Thread.

I know that my WriteHandler somehow got into this state.  As to how it got into this state I don't know.  After learning more about NIO I think my theory above is wrong.  I'll update the issue if I get a better theory and/or a test case.

**Comment by youngm** [comment]
I cannot for the life of me figure out why this selector won't select for this SelectionKey.

sk.isWritable() == true
after it fails to select I confirmed.
selector.selectedKeys().isEmpty() == true

Why else would a selector not select on a SelectionKey waiting to write?  This is killing me!!!!

**Comment by Donnerbart** [comment]
Is it possible to share your setup with us? This would help to reproduce the issue faster, if it cannot be figured out by looking at the code only.

**Comment by youngm** [comment]
So, now I'm looking at this possibly being a java bug.  When invoking `Selector.select()` eventually `sun.nio.ch.EPollSelectorImpl.updateSelectedKeys()` sees that there is 1 update so when it invokes this.pollWrapper.getDescriptor(0) it gets a file descriptor different from the descriptor of the channel in the only SelectionKey in the Selector.

I think I'm in over my head on this one.  I think my next step is to take @Donnerbart 's advice and try and make something reproducible.

**Comment by pveentjer** [comment]
Did you manage to get it reproducible? As soon as we have released Hazelcast 3.5-EA we'll be focussing on this issue and reproducing suchs problems is often half the battle.

**Comment by youngm** [comment]
@pveentjer Thanks for checking in on me.  I'm still trying to get something that can be fairly easily reproduced outside of my test environment.  I'll update the issue if I discover anything new.

**Comment by pveentjer** [comment]
Could you try using this branch:

https://github.com/hazelcast/hazelcast/pull/5196

And add the following system properties:

-Dhazelcast.performance.monitoring.enabled=true -Dhazelcast.performance.monitor.delay.seconds=10

This will dump the new metrics; with a ton of stats regarding the io system. I have added a lot of these metrics just to deal with IO problems like this.

**Comment by youngm** [comment]
@pveentjer will do.  Plus I think I understand the problem well enough to put together a way for your QA to reproduce this given 10 or so VMs.  I'll hopefully get something uploaded in the next week or so.  In my test case build I'll include this version of hazelcast.

**Comment by jerrinot** [comment]
Hi @youngm: Thanks for a really nice work here!
Did you have a chance to run the test with the branch @pveentjer was referring to?

**Comment by youngm** [comment]
@jerrinot No, I've been out of town the last couple of days.  I hope to get back to it later this week.

**Comment by jerrinot** [comment]
great, ping me if you need any assistance!

**Comment by youngm** [comment]
This test case includes the source and a build of a custom memcache server backed by hazlecast (memcache-hazelcast) and a stress test client (memcache-test).

memcache-hazelcast provides a memcache server written in netty.
memcache-test provides stress tests using the xmemcached java client.

The server is an open source project here: https://github.com/cloudfoundry-community/memcache-hazelcast

Useful hazelcast code resides mostly:
* https://github.com/cloudfoundry-community/memcache-hazelcast/blob/master/src/main/java/cloudfoundry/memcache/hazelcast/Main.java#L134
* https://github.com/cloudfoundry-community/memcache-hazelcast/blob/master/src/main/java/cloudfoundry/memcache/hazelcast/HazelcastMemcacheMsgHandlerFactory.java#L37
* https://github.com/cloudfoundry-community/memcache-hazelcast/blob/master/src/main/java/cloudfoundry/memcache/hazelcast/HazelcastMemcacheMsgHandler.java

The code uses IExecutorService extensively to help keep the netty server async.

To duplicate the behaviour experienced in this issue I used the following:

* Oracle build of java 8u40 64bit.
* Instructions to configure and run the test are in the included README
* Ran a 2 node server.  Each on an 8 CPU Ubuntu Trusty VM. (Also duplicated on Ubuntu Lucid haven't tried any other OS)
* Run anywhere from 3-10 client vms.  Also each with 8 cpus running Ubuntu Trusty.

Keep adding client vms and let run for a few minutes until you see clients begin to fail.  I usually see failure happen at about 3-10 clients.  Some seconds later servers will start to show "No response for 120000 ms." errors.

If you don't see failure happen after growing to say 15 clients let it run all night.  I haven't been able to confirm it but it seems once a server has experienced this issue it seems to happen sooner subsequent times without restarting the vm.

Once the client start failing the server will be in its broken state.  One of the servers will not be able to write to the other one until the channel it is using is reconnected.  This might happen as part of a heartbeat so I've increased heartbeat timeout to a larger value (see server.yml)

The memcache-hazelcast build included in the .zip contains a build from pveentjer's monitor branch.  Feel free to rebuild with different hazelcast version if you wish.

I've included performance files from a 2 node test I recently ran while in a broken state.

You can download the test here: https://drive.google.com/file/d/0ByIViWrm6MZERFYxMW9uSE5sZ3c/view?usp=sharing

I hope you are able to duplicate it. :)

**Comment by jerrinot** [comment]
Hi Mike,

Once again thanks for an awesome report and reproducer! I'll try to
reproduce it today or tomorrow.

**Comment by noctarius** [comment]
Yes absolutely amazing! Hope @jerrinot will figure it out but I have high expectations :)

PS: From community aspect, what is the difference between the HZ built-in memcache server and your implementation? Would be nice to get some more information about the project :)

**Comment by jerrinot** [comment]
Hi @youngm 

I'm getting this error when I'm starting the client:
```
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:480)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:118)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:686)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:320)
	at cloudfoundry.memcache.test.Main.main(Main.java:43)
	... 6 more
Caused by: org.springframework.beans.factory.BeanCreationException: Could not autowire field: private net.rubyeye.xmemcached.MemcachedClient cloudfoundry.memcache.test.MainController.mc; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder 'vcap.services.memcache-service.credentials.password' in string value "${vcap.services.memcache-service.credentials.password}"
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:561)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:88)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:331)
	... 20 more
Caused by: java.lang.IllegalArgumentException: Could not resolve placeholder 'vcap.services.memcache-service.credentials.password' in string value "${vcap.services.memcache-service.credentials.password}"
	at org.springframework.util.PropertyPlaceholderHelper.parseStringValue(PropertyPlaceholderHelper.java:174)
	at org.springframework.util.PropertyPlaceholderHelper.replacePlaceholders(PropertyPlaceholderHelper.java:126)
	at org.springframework.core.env.AbstractPropertyResolver.doResolvePlaceholders(AbstractPropertyResolver.java:204)
	at org.springframework.core.env.AbstractPropertyResolver.resolveRequiredPlaceholders(AbstractPropertyResolver.java:178)
```
I believe I'm starting it as expected
```
java -jar memcache-test.jar --config=file:client.yml
```
I changed the `client.yml` accordingly, but it seems it's ignored. 

I don't know Spring Boot that well, but I assume I can just change the addresses in `application.yml` and rebuild the project. Is it correct?

**Comment by youngm** [comment]
@jerrinot Sorry the correct command for the client is actually:

```
java -jar memcache-test.jar --spring.config.location=file:client.yml
```

@jerrinot The main difference is this memcache server is multi-tenant and built to be a service for Cloud Foundry.  You can read more about it here: https://github.com/cloudfoundry-community/memcache-release

**Comment by pveentjer** [comment]
@youngm did you manage to use the metrics branch?

https://github.com/hazelcast/hazelcast/pull/5322

**Comment by youngm** [comment]
@pveentjer yes, I believe the test case I provided is using a build from the metrics branch.

**Comment by youngm** [comment]
@pveentjer I also included some metrics log output in the .zip I provided.

**Comment by pveentjer** [comment]
Ok. great..  We'll have a look at it today. The main focus of the QA team is to get this d*mn OperationTimeoutException fixed.

**Comment by youngm** [comment]
Great!  I really hope you're able to duplicate it.

**Comment by youngm** [comment]
@jerrinot One other thing I noticed but wasn't able to confirm.  Slowing down the server seemed to make the issue a bit harder to duplicate.  For example, running in debug mode it seemed I needed to either add a bit more load or let the test run a little longer.  I also think @pveentjer 's metric build might similarly slow things a little making the issue a bit harder to duplicate.  So, if you have trouble duplicating the issue with the build I gave you try stock 3.4.2 not running debug.

**Comment by youngm** [comment]
@jerrinot any luck duplicating the issue?  Or do I need to keep thinking of ways you might be able to duplicate this?

**Comment by pveentjer** [comment]
Hi Mike, Jaromir and I have been looking at the OperationTimeoutException. And we are confident we have found a cause. We can reproduce it in minutes. But this only is an issue in the 3.5 branch caused by the IO-balancer. We'll probably provide a fix for it today

So we have not (yet) found the cause of why it happens on 3.4.x

**Comment by youngm** [comment]
@pveentjer Great!  I'll watch for that fix and give a 3.5 build another try.  I'll report back if I'm still seeing the issue.

**Comment by youngm** [comment]
@pveentjer & @jerrinot FWIW, I tried a build based on de74fc4cd4cacea08f38eb253acec4b25db8fc7e with my test case and got the same failure I was getting before.  Let me know if I can do anything to help.

**Comment by pveentjer** [comment]
Hi Mike, I think you are pointing to the wrong commit here. 

We have found the cause of a OpTimeoutEx due to a race in the balancer. We have fixed the problem and now in the processing verifying the fix. It has not yet been merged in the master. 

https://github.com/hazelcast/hazelcast/pull/5342

Can you give it a try with this branch? Perhaps you are running into a different cause of the OpTimeoutException and your problem isn't related to the fix we made.

**Comment by pveentjer** [comment]
I want to zoom in to your initial comment

```
he value of selectedKeyCount is ALWAYS 0.

The result is that this WriteHandler never responds to any requests from a certain node (the node associated with the handler).

It is difficult to get into this state so I cannot provide a simple test case duplicating the issue. But I do have a theory of how this could happen.

In WriteHandler.handle():148 the code is not actually registering for a Write (line: 153) until after it has written to the buffer (line: 148). Hence if it were possible for for the channel to process the buffer and flag it as done writing before registerWrite() is invoked then I think the WriteHandler would end up in the state I have described. Never ready to write and the write Selector never saying it has finished writing anything.
```

Perhaps I'm  missing something. If everything was written to the channel and then you call registerWrite than you will be notified by the socket as soon as there is space. Probably you will be immediately notified that there is space, since you were able to write everything. 

So what happens in 3.4 is:

```
- write stuff to socket (nothing is remaining.. you are completely done)
- set ready to false
- register for write
- selector immediately find the selection key for that write handler and calls 'run' since there is space to write in the socket.
- since ready = false, the handle method is called 
- since there is nothing to write, the handle method flips ready to true and exits at line 144.
```

So currently I don't see how your scenario can happen. But this flow is very complicated and it is easy to miss something.

In 3.5 registering for this unneeded write is not done btw.

**Comment by youngm** [comment]
@pveentjer Yeah, I didn't know very much about nio when I wrote that description and that theory there is completely wrong.  I've since done abundant raw nio testing and study and found what you say is true.  Under normal circumstances the selector always selects if it can write.

To try and determine if the bug was JDK Selector related or something else I later ran a test where I modified the write handler to essentially make it blocking instead of async.  I removed all the selector code and make sure I had  one IO thread for each node.  I was able to duplicate the issue in that scenario as well.  When debugging the code I saw the write handler actually write to the channel but the reader would never receive what was written.  Somehow the channel had essentially broke.  And remained that way until closed and reconnected.  That's when when I  decided something lower level must be going on and gave up trying to fix the issue.  I instead filed issue https://github.com/hazelcast/hazelcast/issues/5253 and decided to just create a test case in hopes you could duplicate it and figure it out. :)

Have you or your QA team had any luck duplicating this broken state with the test case I provided using 3.4.2?

**Comment by youngm** [comment]
@pveentjer @jerrinot I tried the branch in https://github.com/hazelcast/hazelcast/pull/5342 as we expected I still fail the same way.  Though now in addition to OperaionTimeoutExceptions I also get log messages like the ones below when things go bad:

```
2015-05-25 16:50:25.598  INFO 6452 --- [vocationsThread] com.hazelcast.spi.OperationService       : [10.61.140.41]:5701 [dev] [3.5-SNAPSHOT] Handled 0 invocation timeouts and 1 backupTimeouts
2015-05-25 16:50:31.600  INFO 6452 --- [vocationsThread] com.hazelcast.spi.OperationService       : [10.61.140.41]:5701 [dev] [3.5-SNAPSHOT] Handled 0 invocation timeouts and 1 backupTimeouts
2015-05-25 16:52:37.621  INFO 6452 --- [vocationsThread] com.hazelcast.spi.OperationService       : [10.61.140.41]:5701 [dev] [3.5-SNAPSHOT] Handled 1 invocation timeouts and 0 backupTimeouts
```

**Comment by pveentjer** [comment]
How reliable can you reproduce the problem and how much time does it take?

**Comment by youngm** [comment]
@pveentjer I can reproduce it 100% in my environment.  I just have to give it load.  As far as how much time and load that is a little trickier.  Once I've experienced the issue I can restart the processes in the cluster and duplicate it again rather quickly 5-15 min.

Occasionally, I'll notice it takes a couple of hours.  One time I switched to new Lucid VMs where before I was using trusty and it ran like a champ for 3 or 4 hours.  I thought I had it fixed and then it broke again.  After that initial break it would only take minutes to break it again after restarting all of the processes in the cluster.  I don't have any explanation for that.

So, usually 5-15 min may take as long as 3-4 hours.  Running a 2 node server with 10 or so clients.

**Comment by jerrinot** [comment]
Hi @youngm: Could you please specify your environment a bit? Thanks!

**Comment by youngm** [comment]
@jerrinot what kind of details do you want?  I mentioned in the test case comment:

* Oracle build of java 8u40 64bit.
* Ran a 2 node server. Each on an 8 CPU Ubuntu Trusty VM. (Also duplicated on Ubuntu Lucid haven't tried any other OS)
* Run anywhere from 3-10 client vms. Also each with 8 cpus running Ubuntu Trusty.

Running on VMWare infrastructure.

**Comment by youngm** [comment]
@jerrinot Are you trying to reproduce this now?  Is there a chat client we can connect on to discuss?

**Comment by jerrinot** [comment]
Let's go to: https://gitter.im/hazelcast/hazelcast

**Comment by youngm** [comment]
Just to update the issue @jerrinot was unable to duplicate the problem with the test case provided in AWS.  So, I think you can de-prioritize this issue.  I'll run some more tests over the next couple of weeks in both amazon and in my infrastructure to try and narrow the problem further.  But, at this point it doesn't appear to be a bug in hazelcast instead some kind of issue with OS config or infrastructure.  You can close this or leave open.  I'll update with more information when I have it.  Thanks tons for you help @pveentjer and @jerrinot 

In the mean time it would be great if https://github.com/hazelcast/hazelcast/issues/5253 could get looked at so at the very least when 2 of my non master nodes experience this issue they can auto reconnect after a time.

Thanks again guys for giving me some time.

**Comment by youngm** [comment]
A quick update to let everyone know I'm still looking into this issue.  I was able to duplicate the issue in AWS using a c3.8xlarge and a significantly larger number of clients.  I choose c3.8xlarge because it had a 10Gb network connection just like our organization's VMs.

I'm going to continue testing with various amounts of load and whatnot with slower network vms and try lowering the network performance of my org VMs.  I'll keep updating the issue as I find more until I think I have something more solid.

**Comment by jerrinot** [comment]
Hi Mike,

Many thanks for the update. May I kindly ask you to try it with the just
released version 3.5 too? There was a bug related to operational scheduling
we fixed in 3.5. I don't think there is a similar bug in the 3.4, but the
scheduling implementation is quite different so it's hard to prove it.

Cheers,
Jaromír

**Comment by youngm** [comment]
@jerrinot certainly.  I'll add that to my list of things to try.

**Comment by jerrinot** [comment]
That's awesome, many thanks and good luck with testing!

**Comment by youngm** [comment]
Confirmed still fails with 3.5 and on CentOS 7 VM.  Still more stuff to try.

**Comment by youngm** [comment]
After 2 days of testing I think I can say with confidence that I don't have this problem running on CentOS 6 in my infrastructure which is interesting.  At least I found an environment I can use that doesn't have this issue.

I wrote a basic nio non-blocking client/server patterned fairly closely to how hazelcast uses selectors, socketchannel, and buffers.  I was unable to duplicate the issue with that basic pure nio test on Ubuntu Trusty in my infrastructure.

Time to spend more effort trying to consistently duplicate the problem in AWS where you guys can take a closer look at it.

**Comment by jerrinot** [comment]
@youngm: So you are able to reproduce it on multiple versions of Ubuntu, CentOS 7, but not on CentOS 6? That's interesting, it's pointing to some very low-level problem. I think @mdogan has experienced something similar while testing on Google Cloud quite some time ago?

Are your OSs fully updated? Do they use the latest available kernel package? I do recall a bug in a kernel causing a thread never woke-up - https://groups.google.com/forum/#!topic/mechanical-sympathy/QbmpZxp6C64

**Comment by youngm** [comment]
@jerrinot One of my todo items is to run my tests against a typical VM install.  As of now all my VMs are based on Cloud Foundry Bosh "Stemcells".

https://bosh.io/stemcells

Which are slimmed down distros that should be fully up to date at the time of building.  To date these stemcells are very well built.  But I agree I should try a stock installed distro to eliminate this variable.  I'll update the issue when I get a chance to try some stock vms.

**Comment by mdogan** [comment]
@jerrinot, right. 

It was on GCE with Ubuntu (12 or 14, I can't remember). Problem was at some point tcp connection between some nodes were becoming stuck, nodes were not able to read/write from/to these sockets. When I inspected the socket send and receive buffers of related connections and it was always like send buffers are full, receive buffers are empty. But data was not flowing anymore. 

Then I switched to Centos 6 and never saw that problem again :)

**Comment by youngm** [comment]
@mdogan @jerrinot That is the EXACT same problem I am seeing here!  Nice to see that I'm not alone. :)

I can confirm the same thing happens in vSphere, Ubuntu 10, and CentOS 7.  If I got something duplicated in AWS would that even be useful to you guys?  Or would you probably just recommend using CentOS 6 and hope whatever is wrong with linux and Java fixes itself eventually?

**Comment by mdogan** [comment]
It will be definitely useful, especially if you can provide exact details to reproduce.

**Comment by bwzhang2011** [comment]
@youngm, any update with such issue ?

**Comment by youngm** [comment]
@bwzhang2011 No, I haven't had time to put together steps to duplicate it in AWS so that Hazelcast can take a closer look.  Now that it is working with CentOS 6 it has become a lower priority for my team.

I was able to duplicate it once using AWS c3.8xlarge instances.  I just need to do it again in some kind of more minimal fashion since those instances are expensive.

**Comment by johmag** [comment]
Hi,

I stumbled upon this issue a few weeks ago. We have experienced what I believe is this same issue when we ran performance test to see if we could migrate from Red Hat 6 to Ubuntu 14. 

What happens is pretty much exactly the same symptoms that are described above by @youngm and @mdogan that during high load after a while (could be a couple of hours or less) connections between some members in the cluster become "stuck" and there is no traffic coming through (the send buffers from one node is full and the receive buffer on the other side empty) and any partition operation will just time out with exceptions like:

2016-03-30 12:13:28,001 [hz.gaming.InspectInvocationsThread] INFO  com.hazelcast.spi.OperationService   - [10.20.1.50]:5305 [xxx] [3.5.3] Handled 0 invocation timeouts and 1 backupTimeouts

and

No response for 120000 ms. Aborting invocation! Invocation{ serviceName='hz:impl:mapService', op=com.hazelcast.map.impl.operation.QueryPartitionOperation{serviceName='null', partitionId=8, callId=2239229, invocationTime=1459341048114, waitTimeout=-1, callTimeout=60000}, partitionId=8, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeout=60000, target=Address[10.20.1.54]:5305, backupsExpected=0, backupsCompleted=0} No response has been received!  backups-expected:0 backups-completed: 0

If the connection that becomes stale is between the master node and a member node, then the cluster will heal after a while as the member is evicted and forced to re-connect which will clear up the connection. However, if it is a connection between to ordinary members it will never self-heal (as described in the other issue #5253 posted by @youngm )

Since we've started to see this we've tested and reproduced this on both Hazelcast versions 3.5.3 & 3.6.1. 

Having seen this issue posted here we decided to test on different OSs and this was the outcome:

* Ubuntu 14 (3.19.0-43-generic #49~14.04.1-Ubuntu) - **Fail**
* CentOS 6 (2.6.32-573.el6.x86_64) - **Works**
* Red Hat 6 (2.6.32-573.el6.x86_64) - **Works**
* Red Hat 7 (3.10.0-327.el7.x86_64) - **Fail**

All these tests were conducted running on VM-Ware & with Oracle JVM 1.8.0_66 or 1.8.0_73.

A possible theory is that there may be an issue that exist with newer versions of the Linux kernel (2.x compared to 3.x), but haven't done enough tests or pin-pointed the issue to confirm that of course.

I am curious to know if there has been any additional findings from you since this original issue was posted that may help us in finding the exact cause. As it looks now we will need to continue to run on RedHat 6 and will not be able to upgrade to 7 (or switch to Ubuntu 14 for that matter).

**Comment by jerrinot** [comment]
@johmag: Thanks for a very thorough info. Do you happen to use Haswell-based CPUs?

**Comment by jerrinot** [comment]
the "the send buffers from one node is full and the receive buffer on the other side empty" is a solid indication there is something funky going on on the OS-level. 

There is a known issue in some kernel versions. See: https://groups.google.com/forum/#!topic/mechanical-sympathy/QbmpZxp6C64 - this Linux bug is known to affect Hazelcast, but I guess it would have had a different manifestation.

**Comment by johmag** [comment]
@jerrinot Yes we are running Haswell-based CPU:s.

**Comment by jerrinot** [comment]
@johmag: what OS/kernel version is the host OS running?

**Comment by johmag** [comment]
Hey @jerrinot if you are referring to the hypervisors we are running VMware ESXi, 6.0.0, 2494585

**Comment by jerrinot** [comment]
@johmag: yep. that's why I meant. Now I am not expert on VMWare, but I assume the hypervisor itself is based in Linux kernel. According to https://esxi-patches.v-front.de/ESXi-6.0.0.html it looks quite old. 

Although the fact that the 2.6 kernel is running fine in guests suggest that the hypervisor itself is not causing it.

**Comment by youngm** [comment]
@johmag I worked really hard on this for a good month and ran out of time.  I even tried to write some simplified test cases using NIO similar to the way Hazelcast does to duplicate the issue outside of Hazelcast and failed.   We decided to just leave it on CentOS 6 for as long as we can

OS level issues are too low for me.  Right now I plan on testing it on Ubuntu 16.04 once it is out and praying whatever was broken is fixed. :)

BTW our environment is the same except we're running vSphere 5.5 and I don't know what type of processor the host is.  I'm pretty certain I was able to duplicate the issue a couple of times on AWS using more expensive 10Gb network instances but that was near the end of my work on the project and I never got a Hazelcast duplicatable test case together.

**Comment by jerrinot** [comment]
I have observed a similar issue while running tests on AWS: a connection was stuck for no obvious reason. 
Also I could see this in syslog: ` xen_netfront: xennet: skb rides the rocket: 20 slots`

Apparently there is a bug in some older versions of Xen driver in Linux kernel. 
Further Sources: 
- https://blog.stathat.com/2014/12/22/fix_ec2_network_issue_skb_rides_the_rocket.html
- http://www.brendangregg.com/blog/2014-09-11/perf-kernel-line-tracing.html
- https://github.com/torvalds/linux/commit/11d3d2a16cc1f05c6ece69a4392e99efb85666a6
- https://github.com/torvalds/linux/commit/97a6d1bb2b658ac85ed88205ccd1ab809899884d

The workaround is to disable scatter-gather: `ethtool -K eth0 sg off`

**Comment by youngm** [comment]
Thanks @jerrinot Next time I run tests I'll give this a try.  I don't know anything about AWS or virtualization.  Any possibility this Xen driver bug could effect VMware as well?  I don't recall seeing any log messages like that when hitting this on vmware.

**Comment by jerrinot** [comment]
@youngm: I don't think it's directly related to the VMWare bug as the two patches are in Xen specific code. Of course there is chance there it a similar problem in VMWare code. 

However VMWare hypervisor is(?) closed source so I'm not sure how to validate it.

**Comment by youngm** [comment]
I'm happy to announce that I can no longer duplicate this issue.  I haven't touched this project for over a year.  So, I spent the last few weeks upgrading everything about my application to latest versions to see if I still see the problem and I don't anymore.  I'm not sure exactly which upgrade fixed the problem because I didn't change them incrementally.  So, I'll list all the changes since my last failed test that I think could have contributed to a fix.

* Upgraded to latest Ubuntu Trusty update with new 4.4 kernel
* Upgraded to openjdk 1.8.0_91
* Upgraded from Hazelcast 3.5 to 3.7
* Since I have a 10GB network I increased my send/receive buffers to 2M each (Should have tried this before but didn't see the properties before)
* I believe our vmware infrastructure was upgraded from 5.5 to 6.0 (not 100% sure)

I've been hammering my application all week continuously with load tests that previously brought on a failure in a matter of an hour or so with no issues.

Because it is possible not everyone who contributed to the issue is fixed I'll let @jerrinot decide if he wishes to close the issue or not.

Thanks,
Mike

**Comment by youngm** [comment]
FWIW, it turns out I thought my trusty vms were upgraded to kernel 4.4   Turns out they are on 3.19.  So perhaps the kernel upgrade had nothing to do with the fix.

**Comment by bwzhang2011** [comment]
@youngm, as 3.8-EA is going to be released. would you mind testing such in your env to verify it once again ?

**Comment by youngm** [comment]
@bwzhang2011 Sure!  I'll give 3.8-EA a try when it is released and report back.  3.7.3  is working great for me BTW.

**Comment by cforce** [comment]
I would try to switch off LRO, also see
https://github.com/hazelcast/hazelcast/issues/8831#issuecomment-328378605

**Comment by youngm** [comment]
Thanks @cforce I haven't had an issue for a while but may do it as a defensive measure if others are seeing issues.
